includes/widgets/class-amp-widgets.php

```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Register widgets, and add actions.
+ *
+ * @package AMP
+ */
+
+/**
+ * Register the widgets.
+ */
+class AMP_Widgets {
+
+	/**
+	 * Add the actions.
+	 *
+	 * @return void.
+	 */
+	public function init() {
+		add_action( 'widgets_init', array( $this, 'register_widgets' ) );
+	}
+
+	/**
+	 * Add the filters.
+	 *
+	 * @return void.
+	 */
+	public function register_widgets() {
+		$widgets = self::get_widgets();
+		foreach ( $widgets as $native_wp_widget => $amp_widget ) {
+			unregister_widget( $native_wp_widget );
+			register_widget( $amp_widget );
+		}
+	}
+
+	/**
+	 * Get the widgets to unregister and register.
+	 *
+	 * @return array $widgets An associative array, with the previous WP widget mapped to the new AMP widget.
+	 */
+	public function get_widgets() {
+		return array(
+			'WP_Widget_Archives'        => 'AMP_Widget_Archives',
+			'WP_Widget_Categories'      => 'AMP_Widget_Categories',
+			'WP_Widget_Media_Audio'     => 'AMP_Widget_Media_Audio',
+			'WP_Widget_Media_Gallery'   => 'AMP_Widget_Media_Gallery',
+			'WP_Widget_Media_Image'     => 'AMP_Widget_Media_Image',
+			'WP_Widget_Media_Video'     => 'AMP_Widget_Media_Video',
```

Comment:
I can see Travis is failing due to these media widgets not existing in 4.7. So I suggest before returning this array to loop over the keys and remove any for which `! class_exists( $key )`.

Reply:
Hi @weston,
That's a good idea. Here's the [class_exists()](https://github.com/Automattic/amp-wp/pull/870/files#diff-4a5a57759183e45d3afe37b6f63e406cR55) check, and the function is slightly refactored.